tezos-encoding/src/nom.rs

```diff
@@ -572,6 +573,56 @@ where
     }
 }
 
+#[derive(Debug, PartialEq)]
+/// Lazy, memoized deserialization of a dynamic block.
+pub struct LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    // Invariant: when the `deserialized` field is some, it's the
+    // result of deserialization of the `serialized` field using
+    // dynamic(T::nom_read)
+    serialized: &'a [u8],
+    deserialized: Option<T>,
+}
+
+impl<'a, T> NomReader<'a> for LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    fn nom_read(input: &'a [u8]) -> NomResult<'a, Self> {
+        map(dynamic(rest), |serialized| LazyCell {
+            serialized,
+            deserialized: None,
+        })(input)
+    }
+}
+
+impl<'a, T> LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    pub fn to_bytes(self) -> &'a [u8] {
```

Comment:
not sure this should be in the public API? The bytes here won't necessarily correspond to just the bytes corresponding to the encoded T

Reply:
It will never correspond, the length prefix from the `dynamic` encoding is missing here. After I move this to types.rs (as you suggested in https://github.com/trilitech/tezedge/pull/46#discussion_r1400585869) I will probably replace this by an `impl<'a> BinWriter for LazyCell<'a>`.